rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,301 @@
+
+### Composition
+
+Using Radix's `asChild` approach, you can compose the `Form` primitive parts with your own components.
+
+```jsx
+<Form.Field name="name">
+  <Form.Label>Full name</Form.Label>
+  <Form.Control asChild>
+    <TextField variant="primary" />
+  </Form.Control>
+</Form.Field>
+```
+
+It can also be used to compose other types of controls, such as a `select`:
+
+```jsx
+<Form.Field name="country">
+  <Form.Label>Country</Form.Label>
+  <Form.Control asChild>
+    <select>
+      <option value="uk">United Kingdom</option>…
+    </select>
+  </Form.Control>
+</Form.Field>
+```
+
+### More on validation
+
+#### Providing your own validation messages
+
+When no `children` are provided, `Form.ClientMessage` will render a default error message for the given `type`.
+
+```jsx
+<ClientMessage type="missingValue" /> // will yield "This value is missing."
+```
+
+You can provide a more meaningful message by providing your own `children`. This also allows for internationalization.
+
+```jsx
+<ClientMessage type="missingValue">Please provide a name.</ClientMessage> // will yield "Please provide a name."
+```
+
+#### Client-side validation types
+
+`Form.ClientMessage` accepts a required `type` prop which is used to determine when the message should show. It matches the native HTML validity state (`ValidityState` on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState)) which result from using attributes such as `required`, `min`, `max`. The message will show if the given `type` is `true` on the control’s validity state.
+
+```ts
+// `type` is one of the following keys from the DOM ValidityState interface:
+
+interface ValidityState {
+  readonly badInput: boolean;
+  readonly customError: boolean;
+  readonly patternMismatch: boolean;
+  readonly rangeOverflow: boolean;
+  readonly rangeUnderflow: boolean;
+  readonly stepMismatch: boolean;
+  readonly tooLong: boolean;
+  readonly tooShort: boolean;
+  readonly typeMismatch: boolean;
+  readonly valid: boolean;
+  readonly valueMissing: boolean;
+}
+```
+
+This means you can even show something whe the field is valid:
+
+```jsx
+<Form.ClientMessage type="valid">✅ Valid!</Form.ClientMessage>
+```
+
+#### Custom validation
+
+On top of all the built-in client-side validation types described above you can also provide your own custom validation whislt still making use of the platform's validation abilities. It uses the `customError` type present in the constraint validity API.
+
+You can pass `type="customError"` on `Form.ClientMessage` and provide an `isValid` prop with your custom validation logic.
+
+> `isValid` is called with the current value of the control and the current values of all other controls in the form.
+
+Here's a contrived example:
+
+```jsx
+<Form.Field name="name">
+  <Form.Label>Full name</Form.Label>
+  <Form.Control />
+  <Form.ClientMessage type="customError" isValid={(value, fields) => value === 'John'}>
```

Comment:
isValid` seems like a boolean prop at first glance, instead of a function. How about `validator`?